File: ocaml/11-20/12.ml
(* Decode a run-length encoded list. (medium)

Given a run-length code list generated as specified in the previous problem, construct its uncompressed version.
*)

type 'a rle =
  | One of 'a
  | Many of int * 'a

let rec toList = function
  | One x -> [x]
  | Many (0, s) -> []
  | Many (n, s) -> s::(toList (Many ((n-1), s)))

let decode xs = List.concat (List.map toList xs)

let _ = assert (
  (decode [Many (4,"a"); One "b"; Many (2,"c"); Many (2,"a"); One "d"; Many (4,"e")]) =
  ["a"; "a"; "a"; "a"; "b"; "c"; "c"; "a"; "a"; "d"; "e"; "e"; "e"; "e"])

File: prolog/11-20/12.pl
/*
Given a run-length code list generated as specified in problem 1.11. Construct its uncompressed version.

decode([[4,a],b,[2,c],[2,a],d,[4,e]], Xs).
Xs = [a, a, a, a, b, c c, a, a, d, e, e, e, e].
*/

dec(_, 0, Acc, Acc).
dec(A, N, Acc, Xs) :- N > 0, N1 is N - 1, dec(A, N1, [A|Acc], Xs).

decode([], Acc, Decs) :- reverse(Acc, Xs), flatten(Xs, Decs).
decode([[N|H] | T], Acc, Decs) :- 
  dec(H, N, [], As),
  decode(T, [As|Acc], Decs).
decode([A|T], Acc, Decs) :- 
  not(is_list(A)),
  decode(T, [[A]|Acc], Decs).

decode(Xs, Decs) :- once(decode(Xs, [], Decs)). 

